c_reference/src/dscnn.c

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include "dscnn.h"
+#include "conv1d.h"
+#include "utils.h"
+
+int phon_pred_lr_cnn(float* output_signal, float* input_signal,
+  unsigned in_time, unsigned in_channels,
+  float* mean, float* var, unsigned affine, float* gamma, float* beta, unsigned in_place,
+  unsigned cnn_hidden, unsigned cnn_padding, unsigned cnn_kernel_size,
+  const void* cnn_params, unsigned cnn_stride, int cnn_activation) {
+  
+  unsigned out_time = in_time - cnn_kernel_size + 2 * cnn_padding + 1;
+  if (in_place) {
+    // BatchNorm
+    batchnorm1d(0, input_signal,
+      in_time, in_channels, 
+      mean, var, affine, gamma, beta,
+      in_place, 0.00001);
+    // CNN
+    conv1d_lr(output_signal, out_time, cnn_hidden, input_signal, 
+      in_time, in_channels, cnn_padding, cnn_kernel_size, 
+      cnn_params, cnn_stride, cnn_activation);
+  }
+  else {
+    // BatchNorm
+    float* norm_out = (float*)malloc(in_time * in_channels * sizeof(float));
+    batchnorm1d(norm_out, input_signal,
+      in_time, in_channels, 
+      mean, var, affine, gamma, beta,
+      in_place, 0.00001);
+    // CNN
+    conv1d_lr(output_signal, out_time, cnn_hidden, norm_out, 
+      in_time, in_channels, cnn_padding, cnn_kernel_size, 
+      cnn_params, cnn_stride, cnn_activation);
+    free(norm_out);
+  }
+  return 0;
+}
+
+int phon_pred_depth_point_lr_cnn(float* output_signal, float* input_signal,
+  unsigned in_time, unsigned in_channels,
+  float* mean, float* var, unsigned affine, float* gamma, float* beta, unsigned in_place,
+  unsigned depth_cnn_hidden, unsigned depth_cnn_padding, unsigned depth_cnn_kernel_size,
+  const void* depth_cnn_params, unsigned depth_cnn_stride, int depth_cnn_activation,
+  unsigned point_cnn_hidden, unsigned point_cnn_padding, unsigned point_cnn_kernel_size,
+  const void* point_cnn_params, unsigned point_cnn_stride, int point_cnn_activation,
+  unsigned pool_padding, unsigned pool_kernel_size, unsigned pool_stride, int pool_activation) {
+  
+  // Activation
+  unsigned out_time;
+  float* act_out= (float*)malloc(in_time * (in_channels >> 1) * sizeof(float));
+  semi_sigmoid_tanh(act_out, input_signal, in_time, in_channels);
+
+  in_channels >>= 1;
+  float* depth_out;
+  if (in_place) {
+    // Norm
+    batchnorm1d(0, act_out,
+      in_time, in_channels, 
+      mean, var, affine, gamma, beta,
+      in_place, 0.00001);
+    // Depth CNN
+    out_time = in_time - depth_cnn_kernel_size + 2 * depth_cnn_padding + 1;
+    depth_out = (float*)malloc(out_time * depth_cnn_hidden * sizeof(float));
```

Comment:
Output Time here (assuming its padded and made equal to the input time) will be of 33 steps (for a 1 second original input data). The hidden dimensions/output channels (equal to input channels) for the Depth CNN will be 200. This will be a total of 6600 slots of float memory (26400) for a 1 sec clip/data